pkg/middleware/payment/middleware.go

```diff
@@ -0,0 +1,206 @@
+	ctx := context.WithValue(r.Context(), PaymentKey, &PaymentInfo{SatoshisPaid: 0})
+	next.ServeHTTP(w, r.WithContext(ctx))
+}
+
+func proceedWithSuccessfulPayment(w http.ResponseWriter, r *http.Request, next http.Handler, paymentInfo *PaymentInfo) {
+	ctx := context.WithValue(r.Context(), PaymentKey, paymentInfo)
+	sendPaymentAcknowledgment(w, paymentInfo)
+	next.ServeHTTP(w, r.WithContext(ctx))
+}
+
+func extractPaymentData(r *http.Request) (*Payment, error) {
+	paymentHeader := r.Header.Get(HeaderPayment)
+	if paymentHeader == "" {
+		return nil, nil
+	}
+
+	var payment Payment
+	if err := json.Unmarshal([]byte(paymentHeader), &payment); err != nil {
+		return nil, fmt.Errorf("invalid payment data format: %w", err)
+	}
+
+	return &payment, nil
+}
+
+func requestPayment(w http.ResponseWriter, r *http.Request, walletInstance wallet.PaymentInterface, price int) {
+	derivationPrefix, err := walletInstance.CreateNonce(r.Context())
+	if err != nil {
+		respondWithError(w, http.StatusInternalServerError, ErrCodePaymentInternal,
+			fmt.Sprintf("Error creating nonce: %s", err.Error()))
+		return
+	}
+
+	terms := NewPaymentTerms(price, derivationPrefix, r.URL.String())
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusPaymentRequired)
+	err = json.NewEncoder(w).Encode(terms)
+	if err != nil {
+		return
+	}
+}
+
+func processPayment(
+	ctx context.Context,
+	walletInstance wallet.PaymentInterface,
+	paymentData *Payment,
+	identityKey string,
+	price int,
+) (*PaymentInfo, error) {
+	valid, err := walletInstance.VerifyNonce(ctx, paymentData.DerivationPrefix)
+	if err != nil {
+		return nil, fmt.Errorf("error verifying nonce: %w", err)
+	}
+
+	if !valid {
+		return nil, errors.New("invalid derivation prefix")
+	}
+
+	result, err := walletInstance.InternalizeAction(ctx, wallet.InternalizeActionArgs{
+		Tx: paymentData.Transaction,
+		Outputs: []wallet.InternalizeOutput{
+			{
+				OutputIndex: 0,
+				Protocol:    "wallet payment",
+				PaymentRemittance: &wallet.PaymentRemittance{
+					DerivationPrefix:  paymentData.DerivationPrefix,
+					DerivationSuffix:  paymentData.DerivationSuffix,
+					SenderIdentityKey: identityKey,
+				},
+			},
+		},
+		Description: "Payment for request",
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("payment processing failed: %w", err)
+	}
+
+	var txid string
+	if len(paymentData.Transaction) >= 4 {
+		txid = fmt.Sprintf("tx-%x", paymentData.Transaction[:4])
+	} else {
+		txid = fmt.Sprintf("tx-%x", paymentData.Transaction)
+	}
+
+	return &PaymentInfo{
+		SatoshisPaid:  price,
+		Accepted:      result.Accepted,
+		Tx:            paymentData.Transaction,
+		TransactionID: txid,
+	}, nil
+}
+
+func sendPaymentAcknowledgment(w http.ResponseWriter, paymentInfo *PaymentInfo) {
+	w.Header().Set(HeaderSatoshisPaid, fmt.Sprintf("%d", paymentInfo.SatoshisPaid))
+}
+
+func respondWithError(w http.ResponseWriter, status int, code, message string, extraData ...map[string]any) {
```

Comment:
this is worth to consider in the future, that maybe adopter would like to have his own error handler and maybe he could pass it in the options and this one would be the default one.

(but it's a thought for another feature/task)